r2/kv/store.go

```diff
@@ -21,76 +21,374 @@
 package kv
 
 import (
+	"fmt"
+
 	"github.com/m3db/m3ctl/r2"
 	"github.com/m3db/m3metrics/rules"
+	"github.com/m3db/m3x/clock"
+	"github.com/m3db/m3x/instrument"
 )
 
 type store struct {
-	cfg StoreConfig
-}
-
-// StoreConfig is all the configuration state needed by a kv service.
-type StoreConfig struct {
-	UpdateHelper rules.RuleSetUpdateHelper
+	iOpts        instrument.Options
+	updateHelper rules.RuleSetUpdateHelper
+	ruleStore    rules.Store
+	clockOpts    clock.Options
 }
 
 // NewStore returns a new service that knows how to talk to a kv backed r2 store.
-func NewStore(cfg StoreConfig) r2.Store {
-	return &store{cfg: cfg}
+func NewStore(iOpts instrument.Options, uh rules.RuleSetUpdateHelper, rs rules.Store, cOpts clock.Options) r2.Store {
+	return &store{iOpts: iOpts, updateHelper: uh, ruleStore: rs, clockOpts: cOpts}
 }
 
-func (s *store) FetchNamespaces() ([]string, int, error) {
-	return nil, 0, nil
+func (s *store) FetchNamespaces() (*rules.NamespacesView, error) {
+	nss, err := s.ruleStore.ReadNamespaces()
+	if err != nil {
+		return nil, err
+	}
+
+	namespaces, err := nss.NamespacesView()
+	if err != nil {
+		return nil, s.handleUpstreamError(err)
+	}
+
+	liveNss := make([]*rules.NamespaceView, 0, len(namespaces.Namespaces))
+	for _, ns := range namespaces.Namespaces {
+		if !ns.Tombstoned {
+			liveNss = append(liveNss, ns)
+		}
+	}
+
+	return &rules.NamespacesView{
+		Namespaces: liveNss,
+		Version:    namespaces.Version,
+	}, nil
 }
 
-func (s *store) CreateNamespace(namespaceID string) (string, error) {
-	return "", nil
+func (s *store) CreateNamespace(namespaceID string, uOpts r2.UpdateOptions) (*rules.NamespaceView, error) {
+	nss, err := s.ruleStore.ReadNamespaces()
```

Comment:
Does `ReadNamespaces` return an error in that case? If it does, seems the `CreateNamespace` method would return an error when the namespace doesn't exist, which is probably not what we want?

Reply:
ReadNamespaces returns all of the available namespaces, it will not error if a single namespace doesn't exist. The error case there is that the top level namespaces path doesn't exist.

CreateNamespace will error if it encounters an error that is not Namespace not found.